pkg/controller/job/job_controller.go

```diff
@@ -1549,13 +1572,150 @@ func getBackoff(queue workqueue.RateLimitingInterface, key interface{}) time.Dur
 	return calculated
 }
 
-// countValidPodsWithFilter returns number of valid pods that pass the filter.
-// Pods are valid if they have a finalizer and, for Indexed Jobs, a valid
-// completion index.
-func countValidPodsWithFilter(job *batch.Job, pods []*v1.Pod, uncounted sets.String, expectedRmFinalizers sets.String, filter func(*v1.Pod) bool) int {
-	result := len(uncounted)
+func getNewBackoffInfo(backoffManager *backoffManager, job *batch.Job, newSucceededPods []*v1.Pod, newFailedPods []*v1.Pod) backoffInfo {
+	backoffInf := backoffManager.getBackoffInfo(string(job.UID))
+	sortByFinishedTime(newSucceededPods)
+	sortByFinishedTime(newFailedPods)
+	newBackoffInfo := backoffInfo{
+		backoffInf.failuresAfterLastSuccess,
+		backoffInf.lastFailureTime,
+	}
+	var i, j int
+	for i < len(newSucceededPods) && j < len(newFailedPods) {
+		successTime := getFinishedTime(newSucceededPods[i])
+		failureTime := getFinishedTime(newFailedPods[j])
+		if successTime.After(failureTime.Time) {
+			newBackoffInfo.failuresAfterLastSuccess = newBackoffInfo.failuresAfterLastSuccess + 1
+			newBackoffInfo.lastFailureTime = &failureTime
+			j = j + 1
+		} else {
+			newBackoffInfo.failuresAfterLastSuccess = 0
+			newBackoffInfo.lastFailureTime = nil
+			i = i + 1
+		}
+	}
+
+	if i < len(newSucceededPods) {
+		newBackoffInfo.failuresAfterLastSuccess = 0
+		newBackoffInfo.lastFailureTime = nil
+	}
+
+	if j < len(newFailedPods) {
+		newBackoffInfo.failuresAfterLastSuccess = newBackoffInfo.failuresAfterLastSuccess + int32(len(newFailedPods)-j)
+		lastFinshedTime := getFinishedTime(newFailedPods[len(newFailedPods)-1])
+		newBackoffInfo.lastFailureTime = &lastFinshedTime
+	}
+
+	return newBackoffInfo
+}
+
+func getInitialBackoffParameters(countedPods []*v1.Pod) (int32, *metav1.Time) {
+	sortByFinishedTime(countedPods)
+	failures := 0
+	var lastFailureTime *metav1.Time
+	for _, pod := range countedPods {
+		if pod.Status.Phase == v1.PodSucceeded {
+			break
+		}
+
+		failures = failures + 1
+		if lastFailureTime == nil {
+			finishedTime := getFinishedTime(pod)
+			lastFailureTime = &finishedTime
+		}
+	}
+
+	return int32(failures), lastFailureTime
+}
+
+func sortByFinishedTime(pods []*v1.Pod) {
+	finishedTimeSortFunction := func(p1, p2 *v1.Pod) bool {
+		p1FinishTime := getFinishedTime(p1)
+		p2FinishTime := getFinishedTime(p2)
+
+		return p1FinishTime.Before(&p2FinishTime)
+	}
+
+	sort.Slice(pods, func(i, j int) bool {
+		return finishedTimeSortFunction(pods[i], pods[j])
+	})
+}
+
+func getFinishedTime(p *v1.Pod) metav1.Time {
+	if p.ObjectMeta.DeletionTimestamp != nil {
```

Comment:
I think we should aim to get more precise results here. The deletionTimestamp will point earlier than actual pod termination due to the pod's graceful period, and other possible latencies. 

I suggest computing is as follows:
1. if the pod is in a terminal phase (this will be the only case once PodDisruptionConditions graduate to GA) compute as below (the max of `FinishedAt` for terminated containers)
2. if the pod is in a non-terminal phase use the DeletionTimestamp

Reply:
Done